governator-core/src/main/java/com/netflix/governator/internal/TypeInspector.java

```diff
@@ -0,0 +1,66 @@
+package com.netflix.governator.internal;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+
+import com.google.common.base.Supplier;
+
+/**
+ * Utility class for class, field and method-based introspection.
+ * 
+ * @author elandau
+ */
+public class TypeInspector {
+    /**
+     * visitor interface for introspection processing
+     */
+    public interface TypeVisitor {
+
+        boolean visit(Field field);
+
+        boolean visit(Method method);
+
+        boolean visit(Class<?> clazz);
+    }
+    
+    public static <R, V extends Supplier<R>&TypeVisitor> R accept(Class<?> type, V visitor) {
+        accept(type, (TypeVisitor)visitor);
+        return visitor.get();
+    }    
+
+    public static void accept(Class<?> type, TypeVisitor visitor) {
+        if (_accept(type, visitor)) {
+            // check these only once at the top level 
+            for (Class<?> iface : type.getInterfaces()) {
+                if (!_accept(iface, visitor)) break;
+            }
+        }
+    }
+
+    private static boolean _accept(Class<?> type, TypeVisitor visitor) {
+        if (type == null) {
+            return false;
+        }        
+        
+        boolean continueVisit = visitor.visit(type);
```

Comment:
To reiterate my concern with using a boolean, it results in cyclomatic complexity which hurts readability and testability.  I think something like this would be better.

```java

if (type == null) {
    return false;
}        
        
if (!visitor.visit(type)) {
    return false;
}

for (final Field field : type.getDeclaredFields()) {
    if (!visitor.visit(field)) {
        return false;
    }
}

for (final Method method : type.getDeclaredMethods()) {
    if (!visitor.visit(method)) {
       return false;
    }
}

return _accept(type.getSuperclass(), visitor);
```